### PR TITLE
Feature/volume motion blur

### PR DESCRIPTION
**Pull request by sergeneren**
#### Summary

Added support for eularian deformation motion blur for volumes. 

#### Change Log 

- Moved the MotionPosition enum from camera to a kernel_types.h to be able to use in kernel 
- Added a new function to *geom_volume.h* to support different up_axis configurations 
- Added velocity_scale field to ccl::Object and kernel object to scale volume motion blur 
- Added an fps field to Camera 

#### Testing 

Please see the hdCycles PR: [https://github.com/tangent-opensource/hdcycles/pull/75](https://github.com/tangent-opensource/hdcycles/pull/75)

### CODE REVIEW

**Comment by dedoardo** [comment]
Hi @sergeneren,

The results look good and the implementation in line with the paper equations (6/7 in Eulerian Motion Blur). Just a couple of questions.

-  Is the behavior for out of bounds lookups specified by the extension type (extend/clip/etc)? The Arnold paper mentions dilatating the volume based on the (pre-filtered) max velocity. I wonder if the quality improvements justify it or not. Maybe this is already currently done during export in which case ignore the question. 
- Have you tried running it inside usdview?

File: src/kernel/geom/geom_volume.h [hunk 1]
```diff
@@ -27,6 +27,24 @@ CCL_NAMESPACE_BEGIN
...
 
+/* Return velocity vector transformed by up_axis and object transform */
+
+ccl_device_inline float3 volume_transform_velocity(KernelGlobals *kg,
+                                                    const ShaderData *sd,
+                                                    float3 P)
+{
+  uint up_axis = kernel_tex_fetch(__objects, sd->object).up_axis;
+
+  if(up_axis == 0){ /* Up axis Z*/
```
Comment (boberfly) [inline src/kernel/geom/geom_volume.h:38]: Feels a bit unfortunate that the axis needs checking in kernel code and introducing a branch. I guess there's no chance of defining the up axis in host code before device upload and embedding it into the transform or something along those lines? Maybe a stupid question from me, I am not sure. :)

File: src/kernel/kernel_types.h [hunk 2]
```diff
@@ -1494,7 +1494,7 @@ typedef struct KernelObject {
 
   bool use_motion_blur;
   float velocity_scale;
-
+  uint up_axis;
```
Comment (boberfly) [inline src/kernel/kernel_types.h:1497]: The padding on the end of this struct perhaps needs one removed now that an uint is introduced? Perhaps the checking could be done outside of kernel code though...

**Review by boberfly — COMMENTED** [review]
Hey @sergeneren just a few things that popped out at me about the kernel code, feel free to correct me though as I am not sure on the gory details, I'm sure @skwerner has a better opinion here. :)

File: src/kernel/kernel_shader.h [hunk 3]
```diff
@@ -1289,6 +1289,47 @@ ccl_device_inline void shader_eval_volume(KernelGlobals *kg,
...
+            const float time = kernel_data.cam.motion_position == MOTION_POSITION_END ?
+                                   (1.0f - kernel_data.cam.shuttertime) + sd->time :
+                                   sd->time;
+            /* Velocity scale: Assume m/s as input, convert to m/(shutter duration).
+               Optional scaling parameter is for artistic control. */
+            const float velocity_scale = kernel_data.cam.inv_fps * kernel_data.cam.shuttertime *
+                                         kernel_tex_fetch(__objects, sd->object).velocity_scale;
+
+            /* Find new sampling position as mentioned in
+             * https://www.arnoldrenderer.com/research/Arnold_TOG2018.pdf */
```
Comment (skwerner) [inline src/kernel/kernel_shader.h:1315]: Thumbs up for leaving a pointer to the paper this is based on.

File: src/kernel/geom/geom_volume.h (hunk 1 above)
Comment (skwerner) [reply src/kernel/geom/geom_volume.h:38]: As discussed in the meeting, we should not be checking the up axis for now and assume that the renderer does not care about the up axis.

**Review by skwerner — CHANGES_REQUESTED** [review]

File: src/kernel/geom/geom_volume.h (hunk 1 above)
Comment (sergeneren) [reply src/kernel/geom/geom_volume.h]: up_axis is now removed from the kernel and render object. Cycles doesn't track the up_axis anymore.

**Review by sergeneren — COMMENTED** [review]
In the latest commit velocity transformation based on up_axis has been removed. I've tested the motion blur without conversion and only depending on the object transform and it is working correctly. 

We will assume that it is the dcc's responsibility to correctly transform the volumes based on their own axis configuration 

volume_with_transform

**Comment by sergeneren** [comment]
@dedoardo thanks for the review,

* Dilating the bounding box of the volume by the max filtered velocity has not been implemented here because based on my experience in Maya project I've seen that filtering had a significant impact on the time to first pixel (mainly because my filtering implementation was single-threaded and did lots of sorting). But during the production, out of many shots there was only two cases that needed dilation in the vel field and it was easily solved by expanding the velocity vdb by some voxel amount before export and filtering was only needed once. So it was a deliberate choice.
* I couldn't yet test it in usdview (having a rez package conflict about usd version and I'm trying to resolve it)

**Comment by dedoardo** [comment]
Thanks for the reply @sergeneren 
- Good to know, if it's happening so rarely I agree it's probably best to resolve the issue during export when needed rather than slow down all the other volumes. I'm guessing it was something with very high velocity causing this?
- Ok, I asked exactly because I was running into similar issues and wasn't sure whether it war related to this PR. It's more likely due to building issues.

**Pull request merged**